server/src/controller/PlaceController.hh

```diff
@@ -0,0 +1,201 @@
+        }
+
+        ENDPOINT("GET", "place/{id}", getOne,
+                 PATH(UInt64, id)) {
+            return createDtoResponse(Status::CODE_200, service_.getOne(id));
+        }
+
+        ENDPOINT_INFO(search) {
+            info->summary = "Search places";
+            info->tags.emplace_back("place-controller");
+
+            info->queryParams["query"].required = false;
+            info->queryParams["limit"].required = false;
+            info->queryParams["offset"].required = false;
+
+            info->addResponse<oatpp::Object<dto::PlacePageDto>>(Status::CODE_200, "application/json");
+            info->addResponse<oatpp::Object<dto::StatusDto>>(Status::CODE_404, "application/json");
+            info->addResponse<oatpp::Object<dto::StatusDto>>(Status::CODE_500, "application/json");
+        }
+
+        ENDPOINT("GET", "place", search,
+                 QUERY(String, query, "query", std::string{}),
+                 QUERY(UInt64, limit, "limit", 20U),
+                 QUERY(UInt64, offset, "offset", uint64_t{0})) {
+            return createDtoResponse(Status::CODE_200, service_.search(query, limit, offset));
+        }
+
+        ENDPOINT_INFO(putOne) {
+            info->summary = "Put one place";
+            info->tags.emplace_back("place-controller");
+            info->addSecurityRequirement("JWT Bearer Auth", {});
+
+            info->addResponse<oatpp::Object<dto::PlaceDto>>(Status::CODE_200, "application/json");
+            info->addResponse<oatpp::Object<dto::StatusDto>>(Status::CODE_400, "application/json");
+            info->addResponse<oatpp::Object<dto::StatusDto>>(Status::CODE_401, "application/json");
+            info->addResponse<oatpp::Object<dto::StatusDto>>(Status::CODE_403, "application/json");
+            info->addResponse<oatpp::Object<dto::StatusDto>>(Status::CODE_500, "application/json");
+        }
+
+        ENDPOINT("PUT", "place", putOne,
+                 AUTHORIZATION(std::shared_ptr<auth::JWT::Payload>, authObject),
+                 BODY_DTO(oatpp::Object<dto::PlaceDto>, dto)) {
+            OATPP_ASSERT_HTTP(
+                    dto->id &&
+                    dto->ownerId &&
+                    dto->address &&
+                    dto->latitude &&
+                    dto->longitude,
+                    Status::CODE_400,
+                    "Required parameter not provided"
+            )
+            try {
+                OATPP_ASSERT_HTTP(
+                        authObject->role == 0 ||
+                        (authObject->userId == service_.getOne(dto->id)->ownerId &&
+                         (dto->ownerId == nullptr || dto->ownerId == authObject->userId)),
+                        Status::CODE_403,
+                        "Cannot create or modify place of another user as a regular user"
+                )
+            } catch (HttpError &error) {
+                if (error.getInfo().status != Status::CODE_404) { throw; }
```

Comment:
What is the reason for that if.

Reply:
It also omits the later checks, which slightly improves performance (not really XD).

Reply:
I think I should even move calling this method before accessing `authObject->userId` (don't need to access the property `ownerId` if the entity doesn't exist)

Reply:
`service_->getOne()` throws 404 the later check `(dto->ownerId == nullptr || dto->ownerId == authObject->userId)` is not asserted. Doesn't this mean that we can create a place for someone else?

Reply:
Oh, that's my mistake, thanks for pointing it out

Reply:
resolving as there's a new one on recent changes